### ISSUE TITLE
nix build & nix develop: variable scope inconsistency

### ISSUE THREAD

**Issue by tg-x** [NixOS/nix#4578]
**Describe the bug**

`nix build` & `nix develop` differ in the way they handle variable scoping

With this [flake.nix](https://github.com/p2pcollab/ocaml-p2p/blob/dda59d135a19ecae549f641780e77b9b9f6444f2/flake.nix), `nix build` succeeds while `nix develop --build` fails due to missing dependencies.
In this example `blip` & `urps` are both the names of flake inputs, and are also package names in `prevOcamlPackages`, which is brought into scope on [line 35](https://github.com/p2pcollab/ocaml-p2p/blob/dda59d135a19ecae549f641780e77b9b9f6444f2/flake.nix#L35).
When they are referenced on [line 59-60](https://github.com/p2pcollab/ocaml-p2p/blob/dda59d135a19ecae549f641780e77b9b9f6444f2/flake.nix#L59), they mean different things in `nix build` & `nix develop`:
- in `nix build` they refer to packages from `prevOcamlPackages`
- in `nix develop` they refer to the flake inputs

Workaround: rename flake inputs

**Expected behavior**

The same behaviour is expected in both cases.
Likely `nix build` is correct in this case that it takes the var closest in scope.

**Steps To Reproduce**

Run `nix build` & `nix develop --build`

**Comment by regnat** [comment]
I don't think the two commands could possibly have different scoping rules because they both use the exact same function for doing the evaluation.

However, the correct scoping is that `blip` and `urbs` do come from the flake inputs, because `with` will never shadow an already bound variable (see in https://nixos.org/manual/nix/stable/#idm140737321917712, “The bindings introduced by with do not shadow bindings introduced by other means”). So I guess the derivation just builds fine without them − I actually tried removing them from the `buildInputs` list and it built without complaining.

**Comment by regnat** [comment]
Just digged a wee-bit deeper because I was curious, and it turns out that `nix develop` fails because for some reason it runs `make` rather than the `buildPhase` that `buildDunePackage` fails. The `Makefile` runs `dune build` which fails because `p2p-poldercast` depends on `blip`. Otoh, `nix build` will run the `buildPhase` specified by `buildDunePackage`, which is (essentially) `dune build -p p2p`. And as the `p2p` package doesn't depend on `blip` or `urbs` it builds fine.

**Comment by tg-x** [comment]
Indeed, the issue is then that `nix build` and `nix develop --build` do different things.
I was expecting `nix develop --build` to build the default package like `nix build`, using `buildDunePackage`
and wasn't expecting it to build everything, that explains why it failed.

Is this how it's supposed to work or is this a bug?

The scoping is indeed correct and the `p2p` package doesn't need those inputs,
the confusion was because of the different behaviour of the two commands.

Thanks for checking!

**Comment by roblabla** [comment]
This is not stale, and still a problem. It seems that, in `nix develop`, running `buildPhase` will always use the stdenv/default builder's buildPhase, instead of the derivation's overridden value. Currently using `nix (Nix) 2.11.0`.
